Add support for STM32WL series Sub-GHz radio (SX1262 based)

fabiobaltieri opened the pull request:
Hi, this adds support for the WL SX1262 based radio.

The radio is baked in the SoC, it's internally wired to a dedicated SPI port but the control signals are through the other SoC units (RCC, PWR, EXTI...) rather than using GPIOs.

The proposed implementation uses a new `compatible` for the ST radio, replaces the control lines GPIOs with ST HAL calls and interrupt.

~~The SPI CS logic would normally not belong to the radio driver, but in this case it doesn't matter since the SPI is internally dedicated, and I think it's nice to keep all the SUBGHZ specific modifications in the same file.~~

For `SX126xGetDio1PinState()` I could not find a matching implementation, but I suspect that that's used for some interrupt related workaround which I suspect may not matter with the WL interrupt implementation, which seems to retrigger until the irq bits are cleared.

Tested on a Seeed LoRa-E5 against the `samples/subsys/lorawan/class_a` (hacked to get random DevNonce support) on The Things Network. That setup has a different RF switch port and SoC variation than the `nucleo_wl55jc`, so it'd be great if someone can test it there (I don't own one).

The other implementation plan I considered was to implement a GPIO compatible driver to implement the WL specific radio signal and interrupt behind the Zephyr GPIO API. That I think would have allowed to run the sx1262 driver unmodified, at the cost of a more complex implementation. Up the the sx1262 driver maintainers to decide whether the extra code path here are acceptable or not.

Let me know what you think.

fabiobaltieri committed:
dts: bindings: sx12xx: add ST stm32wl-subghz-radio node

Add a dts node for st,stm32wl-subghz-radio. The device is integrated in
STM32WL series SoCs, and based on the sx1262, but it does not use any
gpio as all the necessary signals are internally connected to various
SoC units.

To account for that, make the redundant gpio optional in the template
files, but mark them as required in the sx1261 and sx1262 definitions,
to match with what's used in the driver.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri committed:
dts: stm32wl: define a subghz-spi radio node

Add a device node for the subghz radio in stm32wl.dtsi. The radio is
present in all chips of the series, so having the node there with the
common properties simplifies the board dts files.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri committed:
boards: nucleo_wl55jc: add lora support

Enable the subghz radio for the lora samples in nucleo_wl55jc.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri committed:
samples: lora, class_a: allow building on nucleo_wl55jc

Add nucleo_wl55jc to platform_allow for lora send/receive and class_a
samples. The board has a variation of the sx1262 driver, this should
allow to build test that code path.

Fix the existing config as well to specify the platforms on a single
line, as currently only the last entry is considered.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri commented:
Since this does not use the SUBGHZ HAL, that needs to be turned off first in the module. Sent https://github.com/zephyrproject-rtos/hal_stm32/pull/102 for it so that has to be merged first.

erwango commented:
@fabiobaltieri I'm seeking internal help from ST Lora expert in order to prodive a useful feedback.

fabiobaltieri commented:
> @fabiobaltieri I'm seeking internal help from ST Lora expert in order to prodive a useful feedback.

Oh cool, thanks for that! Sorry for messing with the tags, was curious to see if that west trick would work to make buildkite happy... turns out it did. :-)

Mani-Sadhasivam commented on dts/bindings/lora/st,subghz-radio.yaml:
```diff
@@ -0,0 +1,12 @@
+
+properties:
+    interrupts:
+      required: true
```
description?

fabiobaltieri replied:
Yup, added.

Mani-Sadhasivam reviewed (changes requested):
Good to see this coming. PR mostly looks good to me apart from one nitpick.

Mani-Sadhasivam commented:
> The other implementation plan I considered was to implement a GPIO compatible driver to implement the WL specific radio signal and interrupt behind the Zephyr GPIO API. That I think would have allowed to run the sx1262 driver unmodified, at the cost of a more complex implementation. Up the the sx1262 driver maintainers to decide whether the extra code path here are acceptable or not.

I'm happy with the current implementation. It'd be better if we can avoid hacking the sx1262 but the other solution of having a dedicated GPIO driver for radio sounds no good to me.

str4t0m reviewed (commented):
Thanks! A few nits:
- I believe you should add st_subghz_radio also to the BUILD_ASSERT in sx126x.c line 34.
- The gpio device structs for reset, busy and dio1 are not needed for SX126X_STM32WL.
You could exclude them from the sx126x_data struct.
- I'd avoid `#ifndef SX126X_STM32WL` for SX1261 and SX1262 specific code and instead introduce a flag named `SX126X_GPIO`, `SX126X_SIGNAL_GPIO`, or something like that.
This would improve the readability a bit and additionally a third implementation could eventually be added easier in the future.

erwango reviewed (changes requested):
Thanks for this work!

After internal review, it seems that the approach is sane, so no problem on ST side.
Main request would be to change the binding to something more specific.

fabiobaltieri commented:
> Thanks! A few nits:
> 
>     * I believe you should add st_subghz_radio also to the BUILD_ASSERT in sx126x.c line 34.
> 
>     * The gpio device structs for reset, busy and dio1 are not needed for SX126X_STM32WL.
>       You could exclude them from the sx126x_data struct.
> 
>     * I'd avoid `#ifndef SX126X_STM32WL` for SX1261 and SX1262 specific code and instead introduce a flag named `SX126X_GPIO`, `SX126X_SIGNAL_GPIO`, or something like that.
>       This would improve the readability a bit and additionally a third implementation could eventually be added easier in the future.

All done, good points, I didn't like those #ifndef either, code looks a fair bit more understandable now.

fabiobaltieri commented on dts/bindings/lora/st,subghz-radio.yaml:
```diff
@@ -0,0 +1,12 @@
+
+description: ST SUBGHZ Radio
+
+compatible: "st,subghz-radio"
```
Oh good point, renamed.

fabiobaltieri commented:
Updates:
- rebased
- added the interrupt description (reference to how it's described in the SoC user manual)
- added the new node to the build assert
- added a define for SX126X_GPIO
- used the new define to drop the non used device nodes
- reworked all the code path to check for SX126X_GPIO or SX126X_STM32WL
- renamed compatible to st,stm32wl-subghz-radio, left the node name and label property unchanged, keep it short, let me know if you think I should change it as well

Pending:
- update the commit hash in the manifest change (not merged yet)

fabiobaltieri commented:
> > The other implementation plan I considered was to implement a GPIO compatible driver to implement the WL specific radio signal and interrupt behind the Zephyr GPIO API. That I think would have allowed to run the sx1262 driver unmodified, at the cost of a more complex implementation. Up the the sx1262 driver maintainers to decide whether the extra code path here are acceptable or not.
> 
> I'm happy with the current implementation. It'd be better if we can avoid hacking the sx1262 driver but the other solution of having a dedicated GPIO driver for radio sounds no good to me.

Yeah I share that feeling, it's a bit of a compromise, but then this seems to be the accepted pattern in Zephyr. Guess we can always abstract these details out if we see more custom implementations showing up on the market.

mniestroj commented on drivers/lora/sx126x.c:
```diff
@@ -386,9 +407,15 @@ void SX126xIoRfSwitchInit(void)
 void SX126xReset(void)
 {
 	LOG_DBG("Resetting radio");
+#if defined(SX126X_GPIO)
 	gpio_pin_set(dev_data.reset, GPIO_RESET_PIN, 1);
 	k_sleep(K_MSEC(20));
 	gpio_pin_set(dev_data.reset, GPIO_RESET_PIN, 0);
+#elif defined(SX126X_STM32WL)
+	LL_RCC_RF_EnableReset();
+	k_sleep(K_MSEC(20));
+	LL_RCC_RF_DisableReset();
+#endif
```
I don't like those switches scattered all over the driver. I would rather want a variant specific callbacks like:
```
struct sx126x_variant {
	int (*init)(struct sx126x_data *data);
	void (*wakeup)(struct sx126x_data *data);
	void (*reset)(struct sx126x_data *data);
	bool (*is_busy)(struct sx126x_data *data);
	int (*dio1_state)(struct sx126x_data *data);
	void (*irq_mask)(struct sx126x_data *data);
	void (*irq_unmask)(struct sx126x_data *data);
};
```
and then implement a variant with gpios:
```
static int sx126x_gpio_init(struct sx126x_data *data)
{
	gpio_init_callback(&data->dio1_irq_callback,
			   sx126x_dio1_irq_callback, BIT(GPIO_DIO1_PIN));
	if (gpio_add_callback(&data->dio1, &data->dio1_irq_callback) < 0) {
		LOG_ERR("Could not set GPIO callback for DIO1 interrupt.");
		return -EIO;
	}
	
	return 0;
};

struct sx126x_variant sx126x_gpio = {
    .init = sx126x_gpio_init,
    ...
};
```
and variant for stm32wl:
```
static int sx126x_stm32wl_init(struct sx126x_data *data)
{
	IRQ_CONNECT(DT_INST_IRQN(0),
		    DT_INST_IRQ(0, priority),
		    radio_isr, DEVICE_DT_INST_GET(0), 0);
	LL_EXTI_EnableIT_32_63(LL_EXTI_LINE_44);
	irq_enable(DT_INST_IRQN(0));
	
	return 0;
}

struct sx126x_variant sx126x_stm32wl = {
    .init = sx126x_stm32wl_init,
    ...
};
```
possibly each implementation should be in a different file (`sx126x_gpio.c` / `sx126x_standalone.c` and `sx126x_stm32wl.c`), conditionally compiled.

fabiobaltieri replied:
So you are thinking about abstracting the variants. Could give it a stab if that's the consensus but doing the whole struct variant thing seems way overkill for what's left of the different code paths now that I got rid of the NSS bit, and since we only support one radio instance, it would bring overhead and no other advantages.

The current driver is already an abstraction, if we want to split the variant in a separate file I'd much rather do a `sx126x_common.c` for the common function, and then keep sx126x.c for the GPIO driver and do a separate one for the ST specific things.

Frankly I'm not sure that's worth the extra file, Kconfig options etc for the six SX126X_STM32WL checks that are left, but let me know what you think.

mniestroj replied:
It is much better now without NSS control, but I would still like to have "gpio vs stm32wl" differences grouped together. It is hard to follow functions if there is if-else on each step, including doing so in preprocessor. What if we would create such functions:
```
#if defined(SX126X_GPIO)

static int sx126x_variant_init(struct sx126x_data *data)
{
	gpio_init_callback(&data->dio1_irq_callback,
			   sx126x_dio1_irq_callback, BIT(GPIO_DIO1_PIN));
	if (gpio_add_callback(&data->dio1, &data->dio1_irq_callback) < 0) {
		LOG_ERR("Could not set GPIO callback for DIO1 interrupt.");
		return -EIO;
	}
	
	return 0;
};

... other functions for gpio variant

#elif defined(SX126X_STM32WL)

static int sx126x_variant_init(struct sx126x_data *data)
{
	IRQ_CONNECT(DT_INST_IRQN(0),
		    DT_INST_IRQ(0, priority),
		    radio_isr, DEVICE_DT_INST_GET(0), 0);
	LL_EXTI_EnableIT_32_63(LL_EXTI_LINE_44);
	irq_enable(DT_INST_IRQN(0));
	
	return 0;
}

... other functions for stm32wl variant

#endif
```

Then we would meet somewhere in the middle, having variant specific grouped together and have no overhead on the runtime and codesize (functions will most likely inline by compiler).

There is also little work to check how much full abstraction would cost us with my initial idea with `struct sx126x_variant`. Hard to say how much overhead it is without trying it out.

fabiobaltieri replied:
Yeah that sounds good, I added a new patch to shuffle the code around and group few functions together, and now the WL ones are in a single block. On the work handler I kept the IRQ renable line for both variants, it should be a noop on the normal one, so we don't have to have an extra case there. At this point it should be easy to add a new Kconfig option and split it into three files, I'd rather keep them on the same file though since we only have two options... less likely for someone to change one and forget the other.

tagunil commented:
I tend to agree with @mniestroj here. I think it would be better to split the driver into several files and define some variant-specific functions.

fabiobaltieri committed:
drivers: spi_ll_stm32: add SUBGHZSPI support

This adds support for controlling the SUBGHZSPI NSS line in STM32WL
devices. This is a special dedicated SPI port only connected to the
radio device internally, chip select happens through a bit in the PWR
module. Adding a special dt-property to identify the port, it all gets
built out on non-WL devices.

Deduplicate the existing dts bindings in the process, and add the new
one for the special spi with the new property.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri commented on drivers/lora/sx126x.c:
```diff
@@ -427,7 +462,13 @@ void SX126xWakeup(void)
 	};
 
 	LOG_DBG("Sending GET_STATUS");
+#if defined(SX126X_STM32WL)
+	LL_PWR_SelectSUBGHZSPI_NSS();
+#endif
 	ret = spi_write(dev_data.spi, &dev_data.spi_cfg, &tx);
+#if defined(SX126X_STM32WL)
+	LL_PWR_UnselectSUBGHZSPI_NSS();
+#endif
```
Yeah I had a comment in the PR description about it. I wanted to keep the device quirks in a single file, but I wasn't too happy with it either. Moved it out on the SPI driver, it's a bit more convoluted but it's where it should be.

fabiobaltieri commented:
From last round of comment: moved the SPI NSS bit into the SPI driver, @erwango @str4t0m you may want to check that separately, kept it all in the second patch (drivers: spi_ll_stm32: add SUBGHZSPI NSS support ).

erwango commented on dts/bindings/spi/st,stm32-spi-common.yaml:
```diff
@@ -0,0 +1,33 @@
+        Example with NSS Pin
+           <&spi3_sck_pc10 &spi3_miso_pc11 &spi3_mosi_pc12 &spi3_nss_pa15>;
+
+    use-subghzspi-nss:
```
I'd prefer we introduce a dedicated spi binding in which this prop is mandatory and exclusive.
This might add some overhead in this PR but in turn simplify ifdefery in the spi driver.

fabiobaltieri replied:
Alright sure, done. Don't think it simplified the ifdefery at all, but at least it makes the new property disappear if there are no special spi active. The Zephyr handling of multiple compatibles feels a bit brittle though... but then it's build tested so we should be good.

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -367,15 +367,34 @@ static int spi_stm32_shift_frames(SPI_TypeDef *spi, struct spi_stm32_data *data)
-			       int status)
+static void spi_stm32_subghzspi_nss_control(const struct device *dev, bool on)
 {
+#if defined(CONFIG_SOC_SERIES_STM32WLX)
```
I'd prefer using a dedicated compat and avoid series related switches (as much as possible)

fabiobaltieri replied:
Cool, this is now a `#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_spi_subghz)`

aurel32 replied:
Another (untested) option might be to force usage the hardware NSS support (this require selecting SPI_STM32_USE_HW_SS and not defining a NSS pin in DT), that way the NSS pin is implicit and does not require any change to the SPI driver.

fabiobaltieri replied:
Yeah I tried that, does not seem to work unfortunately. Double checked that `LL_SPI_SetNSSMode(spi, LL_SPI_NSS_HARD_OUTPUT);` gets called, it does, but then the device does not respond. Sorry.

erwango reviewed (changes requested):
Some comments related to the new SPI driver impacts

fabiobaltieri committed:
drivers: sx126x: refactor few functions for stm32wl support

Move few functions around in preparation of adding the STM32WL variant
of the driver.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri committed:
drivers: sx126x: add support for the stm32wl chip

Add support for the STM32WL integrated radio, based on the sx1262. The
STM32WL implementation does not use any GPIO, and the signals are routed
to other units of the SoC and accessed with the ST HAL.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri commented:
Alright new round of updates:

- SPI:
  - switched form dt property to different compatible
  - moved spi_context_cs_control to the same function as the new subghz specific one (felt silly to pair the two calls manually)
- sx126x:
  - added a new patch to refactors the existing functions around in preparation of the one adding the WL specific ones
  - regrouped the WL ones together, it's now a single block

As I mentioned in the patch comment, subtle change: in this version the irq enable and disable functions are enabled unconditionally. They may be redundant in few cases but I left them for the sake of having less cases.

tagunil commented on drivers/lora/sx126x.c:
```diff
@@ -20,15 +20,22 @@ LOG_MODULE_REGISTER(sx126x, CONFIG_LORA_LOG_LEVEL);
 #if DT_HAS_COMPAT_STATUS_OKAY(semtech_sx1261)
 #define DT_DRV_COMPAT semtech_sx1261
 #define SX126X_DEVICE_ID SX1261
+#define SX126X_GPIO
```
What do you think about using word "external" or "standalone" or something like that instead of "GPIO"?

fabiobaltieri replied:
Yeah fine with either. Standalone? Pick one...

fabiobaltieri replied:
Renamed to SX126X_STANDALONE.

fabiobaltieri commented:
Renamed the SX126X define and rebased. CI is happy now.

mniestroj commented on west.yml:
```diff
@@ -82,7 +82,7 @@ manifest:
       path: modules/hal/st
     - name: hal_stm32
-      revision: dc83915685a5b32a299388e54e0fd259ded86bc4
+      revision: refs/pull/102/head # TODO: change to the committed hash
```
That should be the first commit I guess, to make this patch series bisectable.

fabiobaltieri replied:
Yeah totally, waiting for that to get merged to change the hash and move it.

mniestroj commented on samples/drivers/lora/receive/sample.yaml:
```diff
@@ -7,3 +7,4 @@ tests:
    sample.driver.lora.receive:
      platform_allow: 96b_wistrio
      platform_allow: rak4631_nrf52840
+     platform_allow: nucleo_wl55jc
```
In most cases there is one `platform_allow` entry with multiple space separated platforms. Is the method used here (one `platform_allow` entry for each platform) work correctly?

fabiobaltieri replied:
Hey you are right, it does seem like in only ran for the last one! The files changed had multiple entries already, assumed it was fine. I'll fix those.

fabiobaltieri replied:
That worked, current buildkite run has all the platforms, thanks for spotting the problem. Had a quick look at the other files around the repository and could not find any other instance of this, so we should be good now.

mniestroj reviewed (approved):
Really good job, thanks for contribution!

mniestroj commented:
Added DNM, as we wait for https://github.com/zephyrproject-rtos/hal_stm32/pull/102.

erwango commented:
> [zephyrproject-rtos/hal_stm32#102](https://github.com/zephyrproject-rtos/hal_stm32/pull/102).

Which would need one more approval

mniestroj commented:
>>zephyrproject-rtos/hal_stm32#102.
>
>Which would need one more approval

Do west modules require 3 approvals? Or is it hal_stm32 specific policy?

erwango commented:
> > > [zephyrproject-rtos/hal_stm32#102](https://github.com/zephyrproject-rtos/hal_stm32/pull/102).
> > 
> > 
> > Which would need one more approval
> 
> Do west modules require 3 approvals? Or is it hal_stm32 specific policy?

I would be able to merge it myself, but then I'd need one more approval.

fabiobaltieri commented:
Cool! Rebased, updated the hash in the manifest and moved the manifest patch ahead of all the others.

fabiobaltieri committed:
boards: lora_e5_dev_board: add lora support

Define the subghz radio node in the module dtsi file, with the internal
RF switch connection and TCXO setting, and point at it from the
corresponding dev board file.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

andysan commented on dts/bindings/spi/st,stm32-spi-common.yaml:
```diff
@@ -0,0 +1,26 @@
+# Copyright (c) 2021 Fabio Baltieri
```
This is moved from another source file and should therefore retain the original copyright notice.

fabiobaltieri replied:
Fair point, changed.

andysan commented on drivers/lora/sx126x.c:
```diff
@@ -358,6 +371,60 @@ static int sx126x_irq_init(void)
 	return 0;
 }
 
+#elif defined(SX126X_STM32WL)
```
I would suggest moving these to a separate C file (`sx126x_standalone.c` and `sx126x_stm32wl.c`) and let the build system select the right implementation. That avoids clutter in this file and scales better in case there are more SoCs with similar requirements.

fabiobaltieri replied:
I generally share the feeling, however at this stage, with only one compact alternative implementation I'd rather keep the whole driver it in a single file, and split it once and if a more special versions with different quirks shows up on the market. Unless you are aware of that happening already...

Mani-Sadhasivam replied:
I think it makes sense to do it now itself as it makes the driver look cleaner.

fabiobaltieri replied:
Alright, moved few things around, see my notes in the comment after the push.

andysan reviewed (dismissed):
The copyright header needs to accurately reflect the original copyright of the moved code before merging this change. Moving the device-specific (stand-alone vs SoC) code to separate source files would be a nice-to-have, but isn't a requirement.

str4t0m reviewed (commented):
The commit: "drivers: sx126x: add support for the stm32wl chip" still includes a comment about chip select logic in its commit message, which was now moved to the spi driver. You could remove that comment.
Besides that, only the maintainers decision whether to move the stm32wl specific code into a separate file, or leave it as it is should be left.

Mani-Sadhasivam commented on boards/arm/nucleo_wl55jc/nucleo_wl55jc.dts:
```diff
@@ -101,6 +102,17 @@
 
+&subghzspi {
+	status = "okay";
+	lora: subghz-radio@0 {
```
Node names should be generic. So in this case, it should be `radio@0` as we have a single radio at `0` index. By mistake we added node names based on device names like `sx1276`. I'll get them fixed.

fabiobaltieri replied:
Cool will change that... how about the label property? Should that go to "RADIO" or something else?

fabiobaltieri committed:
drivers: sx126x_standalone: convert to gpio_dt_spec APIs

Convert the standalone part of the sx126x driver to the new gpio_dt_spec
APIs. This allows moving the specific GPIOs in that part of the driver
and out of struct sx126x_data.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri commented:
Alright, updates:
- reworded one outdated commit still mentioning the old NSS logic
- renamed the radio dts node
- split the driver in two

About the split, I would have normally reshuffled the change in the two existing commits: refactor and add, but since there were few approval already (@erwango @mniestroj) I've figured I'll keep the split on a separate commit and give you a chance to review it separately. If you prefer, I can squash it into the refactor patch, and then keep the "add..." one just for the new file related changes.

So the way I went was to add a `sx126x_common.h` with few common function declarations, and called those from the original ones. Went the wrapper way so that we don't have to share the `dev_data` pointer and we can keep it static (which it was not). That gave a chance to also keep some of the logic in the common driver (sleeps etc...).

I also added a patch to isolate the standalone specific gpios to the standalone file, and converted those to the new `gpio_dt_spec` api, left the existing ones untouched for now.

Tested on a WLE5 but only built tested for the standalone variant (I don't have a board for that).

fabiobaltieri commented:
Well scrape the comment about the extra commit, could not stand the previous commit sequence narrative. Refactored the split in three commits: refactoring the existing code out, convert the gpio api, add wl32 support. Sorry about the extra re-reviewing work. Also moved few functions back to their original place to minimize the delta and built test the intermediate patches. Let me know what you think.

fabiobaltieri commented:
Small update:
- rebased
- dropped the HAL patch (it's already being updated by another commit)
- added the subghz node definition for the newly introduced `lora_e5_dev_board` board as well, in the module dtsi file

mniestroj reviewed (approved):
I like it even more now!

tagunil commented on drivers/lora/Kconfig.sx12xx:
```diff
@@ -42,6 +44,12 @@ config LORA_SX126X
 	help
 	  Enable LoRa driver for Semtech SX1261 and SX1262.
 
+config LORA_STM32WL_SUBGHZ_RADIO
```
Wouldn't LORA_STM32WL be enough as a symbol name? It isn't really important, of course.

fabiobaltieri replied:
Yeah that was what I originally picked as well but it felt wrong to just have the platform name, then I added SUBGHZ as that's how it's referred in the manual, but then I realize that that actually identifies the internal SPI port, and ended up here. I'm fine with anything really... @erwango is there a convention we should follow here?

erwango replied:
Unfortunately, I don't think we have any convention on this

str4t0m reviewed (approved):
Thanks for additionally adding the subghz node definitions to the lora-e5 module.
I can confirm that basic lora communication works on the lora_e5_dev_board as well as the nucleo_wl55jc.

erwango commented:
@Mani-Sadhasivam, @andysan Can you have a new look ?

Mani-Sadhasivam commented on drivers/lora/sx126x_common.h:
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2021 Fabio Baltieri
```
nit: sort copyrights based on years.

fabiobaltieri replied:
woops, sorted.

Mani-Sadhasivam reviewed (approved):
LGTM

andysan commented on drivers/lora/sx126x_common.h:
```diff
@@ -0,0 +1,74 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
This file should presumably needs include guards.

fabiobaltieri replied:
Oh yeah, added now.

andysan reviewed (commented):
Thanks for the updated revision. It seems like you forgot include guards in the new header file. It looks good other than that.

fabiobaltieri commented:
Cool, hey thanks everyone for the reviews and comments!

erwango commented:
> Cool, hey thanks everyone for the reviews and comments!

Thanks for your patience and going into these multiple iterations!

JordanYates commented:
It looks like this was merged without even cursory testing of the standalone code. The `struct sx126x_data` instances used in `sx126x.c` and `sx126x_standalone.c` are completely separate variables, resulting in hard faults as soon as any transmission completes when the `sx126x_standalone.c` file tries to queue an uninitialized work item:
https://github.com/zephyrproject-rtos/zephyr/blob/f3435b75efbfd7614ff873f8bfa90a031d6d27e8/drivers/lora/sx126x_standalone.c#L65
https://github.com/zephyrproject-rtos/zephyr/blob/f3435b75efbfd7614ff873f8bfa90a031d6d27e8/drivers/lora/sx126x.c#L42
@Mani-Sadhasivam

fabiobaltieri commented:
> It looks like this was merged without even cursory testing of the standalone code.

Yeah unfortunately I don't own every possible piece of hardware supported by Zephyr, but hey I'm working on it. :-)

> The `struct sx126x_data` instances used in `sx126x.c` and `sx126x_standalone.c` are completely separate variables, resulting in hard faults as soon as any transmission completes as `sx126x_standalone.c` tries to queue an uninitialized work item:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/f3435b75efbfd7614ff873f8bfa90a031d6d27e8/drivers/lora/sx126x_standalone.c#L65
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/f3435b75efbfd7614ff873f8bfa90a031d6d27e8/drivers/lora/sx126x.c#L42

That is not the problem, the `dev_data` pointer in standalone is just a local pointer that refers to the main one. The problem is that the gpio callback device is not the LORA device, it's the GPIO one.

```
Breakpoint 1, sx126x_dio1_irq_callback (dev=0xf2a4 <__device_dts_ord_10>, cb=0x20000604 <dev_data>, pins=8192) at /home/fabiobaltieri/zephyrproject/zephyr/drivers/lora/sx126x_standalone.c:59
59              if (pins & BIT(sx126x_gpio_dio1.pin)) {
(gdb) p dev->data
$1 = (void * const) 0x2000134c <gpio_nrfx_p0_data>
```

Right thing to do is to get back the structure from `cb`, sent a PR to do that.

JordanYates commented:
> Yeah unfortunately I don't own every possible piece of hardware supported by Zephyr, but hey I'm working on it. :-)

This wasn't a dig at you, the maintainers who approved should have a board between them :)

> That is not the problem, the dev_data pointer in standalone is just a local pointer that refers to the main one. The problem is that the gpio callback device is not the LORA device, it's the GPIO one.

True, I always get thrown off by the non-standard way that struct is defined and used...
